From 48be4ffac912e8aad5adb0b878d7bc88a424eca8 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Sun, 19 Feb 2017 17:26:51 +0200
Subject: [PATCH 0834/1345] fix: net: mvpp2x: fix RX CRC errors during ethtool
 port mode transition

commit  0e5e5156e5721f5bfd39501d28be70d189dea966 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Issue:
- CRC error in RX during port mode transition with enabled port.

Issue root cause:
- Calling Serdes PHY ON procedure with enabled PCS clock.

Fix:
- Disable GoP port before calling Serdes PHY ON procedure.
- Change port up order in ethtool set settings callback:
  Serdes PHY ON -> GoP port ON -> netif_carrier_on

Change-Id: I2b39911c5ed691cf5e656d902a33d1d4415b8288
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37131
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c  |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
index a7bfc88..36c3ad5 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
@@ -642,9 +642,11 @@ static int mv_pp2x_ethtool_set_settings(struct net_device *dev,
 		mv_pp22_set_net_comp(port->priv);
 
 		if (mac->flags & MV_EMAC_F_PORT_UP) {
+			mv_gop110_port_disable(gop, mac, port->comphy);
+			phy_power_on(port->comphy);
 			mv_gop110_port_events_unmask(gop, mac);
 			mv_gop110_port_enable(gop, mac, port->comphy);
-			phy_power_on(port->comphy);
+			netif_carrier_on(port->dev);
 		}
 	}
 
-- 
1.7.9.5

